File: data-structures/queue/queue.cpp
#include "queue.h"
#include <string>
#include <vector>

using std::string;
using std::vector;

Queue::Queue(int s) : r_index{0}, w_index{0},bucket{vector<int> (s + 1)} {}
void Queue::enqueue(int elem){
    if (is_full()) throw(QueueException("Full queue"));
    bucket[w_index] = elem;
    w_index = (w_index + 1)%bucket.size();
}

int Queue::front() const {
    if (is_empty()) throw(QueueException("Empty queue"));
    return bucket[r_index];
}

int Queue::dequeue(){
    if (is_empty()) throw(QueueException("Empty queue"));
    int front = bucket[r_index];
    r_index = (r_index + 1)%bucket.size();
    return front;
}

bool Queue::is_empty() const {
    return r_index == w_index;
}

bool Queue::is_full() const {
    return (w_index + 1)%bucket.size() == r_index;
}

File: revision/29_jun_24/src/linear_ds/queue.rs
struct Queue {
    inner: Vec<i32>,
    read: usize,
    write: usize,
}

impl Queue {
    fn new(capacity: usize) -> Self {
        Self {
            inner: vec![0; capacity + 1],
            read: 0,
            write: 0,
        }
    }

    fn is_empty(&self) -> bool {
        self.read == self.write
    }

    fn push(&mut self, elem: i32) {
        let cap = self.inner.len();
        if (self.write + 1) % cap == self.read {
            panic!("cannot push to a full queue")
        }
        self.inner[self.write] = elem;
        self.write = (self.write + 1) % cap;
    }

    fn pop(&mut self) {
        if self.is_empty() {
            panic!("empty queue");
        }
        self.read = (self.read + 1) % self.inner.len();
    }

    fn front(&self) -> i32 {
        if self.is_empty() {
            panic!("empty queue");
        }
        self.inner[self.read]
    }
}

#[cfg(test)]
mod tests {
    use super::Queue;

    #[test]
    fn queue() {
        const CAPACITY: usize = 10;

        let mut queue = Queue::new(CAPACITY);
        assert!(queue.is_empty());
        queue.push(1);
        assert!(!queue.is_empty());
        assert_eq!(queue.front(), 1);
        queue.push(2);
        assert_eq!(queue.front(), 1);

        queue.pop();
        assert_eq!(queue.front(), 2);
    }

    #[test]
    #[should_panic(expected = "cannot push to a full queue")]
    fn queue_push() {
        const CAPACITY: usize = 1;
        let mut queue = Queue::new(CAPACITY);
        queue.push(1);
        queue.push(2);
    }

    #[test]
    #[should_panic(expected = "empty queue")]
    fn queue_pop() {
        const CAPACITY: usize = 1;
        let mut queue = Queue::new(CAPACITY);
        queue.pop();
    }

    #[test]
    #[should_panic(expected = "empty queue")]
    fn queue_front() {
        const CAPACITY: usize = 1;
        let queue = Queue::new(CAPACITY);
        queue.front();
    }
}
